addons/mrp/models/mrp_production.py

```diff
@@ -473,7 +473,18 @@ def _compute_picking_ids(self):
                 ('group_id', '=', order.procurement_group_id.id), ('group_id', '!=', False),
             ])
             order.picking_ids |= order.move_raw_ids.move_orig_ids.picking_id
-            order.delivery_count = len(order.picking_ids)
+            # Count only deliveries related to the current MO, excluding canceled pre-pickings in a 2-step or 3-step MRP process.
+            if order.warehouse_id.manufacture_steps in ['pbm_sam', 'pbm']:
+                filtered_picking_ids = order.picking_ids.filtered(
+                    lambda picking: picking.origin == order.name or any(
+                        order in move.move_dest_ids.raw_material_production_id or
+                        order in move.move_orig_ids.production_id
+                        for move in picking.move_ids
+                    )
+                )
+                order.delivery_count = len(filtered_picking_ids)
```

Comment:
Instead of filtering everywhere, I would set it and compute in direct picking_ids, It'll reduce code redundancy